lib/phoenix_live_view/tag_engine.ex

```diff
@@ -1160,6 +1182,104 @@ defmodule Phoenix.LiveView.TagEngine do
+    {lhs, variables} = mark_variables_as_change_tracked(lhs)
+    for_expr = {:<-, for_meta, [lhs, rhs]}
+
+    # now we build the new ast that we pass to the engine
+    ast =
+      quote do
+        Phoenix.LiveView.TagEngine.keyed_comprehension(
+          {unquote(state.caller.module), unquote(tag_meta.line), unquote(tag_meta.column),
+           unquote(key_expr)},
+          %{unquote_splicing(variables)},
+          do: unquote(invoke_subengine(state, :handle_end, []))
+        )
+      end
+
+    state = pop_substate_from_stack(state)
+
+    keyed_ast =
+      state
+      |> push_substate_to_stack()
+      |> update_subengine(:handle_begin, [])
+      |> update_subengine(:handle_expr, ["=", ast])
+      |> invoke_subengine(:handle_end, [])
+
+    {for_expr, keyed_ast}
+  end
+
+  defp maybe_keyed(state, _type, %{for: for_expr}) do
+    {for_expr, invoke_subengine(state, :handle_end, [])}
+  end
+
+  defp mark_variables_as_change_tracked(ast) do
+    # we traverse the AST and mark elements that should not considered
+    # to be variables as "skip";
+    # we need to handle both the pre and post phases to properly unmark
+    # the skip after we handled an expression
+    {ast, {_, vars}} =
+      Macro.traverse(
+        ast,
+        # we start in collect mode (true) and an empty list of vars
+        {true, []},
+        fn
+          # skip pinned expressions, marking the pin expression and setting
+          # collect to false
+          {:^, meta, [expr]}, {true, vars} ->
+            {{:^, Keyword.put(meta, :skip_mark, true), [expr]}, {false, vars}}
+
+          # skip the right hand side in something like <<foo::binary>>
+          {:"::", meta, [left, right]}, {true, vars} ->
+            # we need to return a list for prewalk to walk the left hand side
+            {new_left, inner_vars} = mark_variables_as_change_tracked(left)
+
+            {{:"::", Keyword.put(meta, :skip_mark, true), [new_left, right]},
+             {false, vars ++ inner_vars}}
+
+          # collect mode -> mark the variable as change_track and collect
+          # it for passing it to the KeyedComprehension component
+          {name, meta, context}, {true, vars}
+          when is_atom(name) and is_list(meta) and is_atom(context) ->
+            var = {name, Keyword.put(meta, :change_track, true), context}
+            {var, {true, [{name, var} | vars]}}
+
+          other, acc ->
+            {other, acc}
+        end,
+        fn
+          {op, meta, args}, {false, vars} when is_atom(op) and is_list(meta) and is_list(args) ->
+            case Keyword.pop(meta, :skip_mark) do
+              {nil, meta} ->
+                {{op, meta, args}, {false, vars}}
+
+              {true, meta} ->
+                # reset the collect mode
+                {{op, meta, args}, {true, vars}}
+            end
+
+          other, acc ->
+            {other, acc}
+        end
+      )
+
+    {ast, vars}
+  end
```

Comment:
This is definitely way more complex than it needs to be. Doing AST recursion would do the job but I think we should solve it instead in `engine.ex`. The engine already skips the right-hand side of `::`, so no variable there would be tracked, and we can add a clause that checks explicitly for `^`, like this:

```elixir
def analyze({:^, _, [{var, _, context}]}, ...) do

end
```

that does not include them on change tracking. This way this code block can be "dumb" and annotate all vars, trusting the engine will handle it accordingly (which it already mostly does).